### PR TITLE
 Implement the nearby_wetdry approach from rainlink 

### PR DESCRIPTION
**Pull request by maxmargraf**
The nearby_wetdry approach is based in the description from [Overeem et al. 2016](https://amt.copernicus.org/articles/9/2425/2016/amt-9-2425-2016.pdf). The implemented method (`nearby_wetdry()`) features default values used in this publication. The classification function is complemented by following methods 
* calculate the distances between CML nodes (`calc_distance_between_cml_endpoints()`) 
* derive "min-max data" from instantaneous sample CML data (`instanteanous_to_minmax_data()`)  
* 
Additionally, the [haversine function](https://github.com/pycomlink/pycomlink/blob/94ee2efdf8fe95a6312a92517fb48efa4fd11b4a/pycomlink/spatial/helper.py#L17) was updated using numpy instead of math functions because numpy can broadcast arrays.  

An example notebook (to be made) will show the usage of the method.

### CODE REVIEW

**Comment by maxmargraf** [comment]
Issues to be fixed:

- [ ] when CML length > r (radius for which CMLs are considered neighboring) this CML has to be included anyway
- [ ] add example notebook

**Comment by cchwala** [comment]
Nice work @maxmargraf 👏 

Can you add an example time series plot which shows the original 1-minute data (from our example data) and the derived 15-minute min-max data. This would be good to give people an idea how "reasonable" our derived 15-minute min-max data is.

**Comment by cchwala** [comment]
If you update the notebook once more, you could also add the CML time series (only Pmin?) to the plots (as a new subplots?) that show the resulting wet-dry and reference rain rate. That would make it clearer what is going on regarding wet-dry.

**Comment by maxmargraf** [comment]
Example time series of Pmin (and other min-max processing variables) as well as raw rsl and tsl were added to the notebook. From my side, this is ready to be merged.

**Comment by cchwala** [comment]
are my two comments regarding specific parts in the code also resolved?

**Review by cchwala — CHANGES_REQUESTED** [review]
Sorry for the confusion. My fault. I did not press "submit review"...

File: pycomlink/processing/wet_dry/nearby_wetdry.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        # only make wet dry detection if min_links is reachied within r
+        if sum(ds_dist.within_r.sel(cml_id1=cmlid).values) > min_links:
+            # select all CMLs within r
+            ds_nearby_cmls = ds_cml.isel(
+                cml_id=ds_dist.within_r.sel(cml_id1=cmlid).values
+            )
+
+            # add selected cml if its excluded because its longer then r
+            if cmlid.values not in ds_nearby_cmls.cml_id:
+                ds_nearby_cmls = xr.concat(
```
[the remark this answers is not in the document]
Comment (maxmargraf) [reply pycomlink/processing/wet_dry/nearby_wetdry.py:178]: I added another synthetic CML to the test which is longer than r. This should now test these lines.

[the remark this answers is not in the document]
Comment (cchwala) [reply pycomlink/processing/wet_dry/nearby_wetdry.py:178]: looks like it does not yet
https://app.codecov.io/gh/pycomlink/pycomlink/pull/113#e732c47fdf91bbaf9a783cbd2642a237-R178

[the remark this answers is not in the document]
Comment (maxmargraf) [reply pycomlink/processing/wet_dry/nearby_wetdry.py]: Thanks for the persistence. I found that I already dealt with the exception this if-clause was meant for in thee [function which computes the distances](https://github.com/maxmargraf/pycomlink/blob/cf8f3e87fbdf532fc4d205d1de26cc6b1db63b72/pycomlink/processing/wet_dry/nearby_wetdry.py#L72). In my local testing regime, I used an old version of the function without these lines. Therefore, locally, the if-clause was needed and therefore tested. 
Long story short, test coverage is up to 100% without changing the test itself.

[the remark this answers is not in the document]
Comment (cchwala) [reply pycomlink/processing/wet_dry/nearby_wetdry.py]: 🎉

File: pycomlink/processing/wet_dry/nearby_wetdry.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+        )
+        ds.loc[dict(cml_id1=cmlid)]["b_to_all_b"][:] = spatial.haversine(
+            np.array(site_b_longitude[i]),
+            np.array(site_b_latitude[i]),
+            np.array(site_b_longitude),
+            np.array(site_b_latitude),
+        )
+    # set distance between endpoints of one cml to 0km to not be rejected when
+    # radius r for the nearby approach is larger than r
+    ds = xr.where(
```
Comment (cchwala) [inline pycomlink/processing/wet_dry/nearby_wetdry.py:72]: From the comment above I do not fully understand what this does. Does it put zeros along the diagonal of the 2D distance matrix?

Is this required because we want zero also for distance between site_a and site_b of the same CML, right?

Maybe you can make this clearer in the comment here and also in the docstring of the function because, even though it is intuitive that the diagonal is set to zero, it is not explicitly clear that this is done and why.

Comment (maxmargraf) [reply pycomlink/processing/wet_dry/nearby_wetdry.py:79]: Yes, your assumption is true.  As result, the distance between site_a and site_b of the same CML is set to 0 km.
I have updated the docstring and the comment in the before the actual command doing this.

**Comment by cchwala** [comment]
Thanks for this contribution 👍 

I will merge it now!
